File: hooks/lazydev-shell-guard.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { platformPaths } from '../runtime/platform-policy.mjs';

function norm(p) { return path.resolve(String(p || '')); }
function artifactDir() { return process.env.LAZYDEV_ARTIFACT_DIR ? norm(process.env.LAZYDEV_ARTIFACT_DIR) : norm(platformPaths().artifactDirectory); }
function loadPrompt() {
  const home = process.env.KIMI_CODE_HOME || path.join(process.env.HOME || process.cwd(), '.kimi-code');
  try { return JSON.parse(fs.readFileSync(path.join(home, 'lazydev-last-prompt.json'), 'utf8')); } catch { return {}; }
}
function artifactIntent(p) { return /(save|simpan|export|download|generate|buat(?:kan)?|create|produce|file|artifact|deliverable)/iu.test(p) && !/(edit|modify|refactor|fix|implement|source|repo|repository|module|component|inside project|in the project|di repo|di project)/iu.test(p); }
function hasWorkspaceWrite(command, cwd) {
  const q=String(command||'');
  if (!q) return false;
  const patterns=[
    /(?:>|>>|tee|cat\s+[^|;&]*>)[\s]*["']?([^\s"'|;&]+\.(?:html?|pdf|docx?|xlsx?|pptx?|zip|png|jpe?g|webp|gif|svg|csv|md|txt))["']?/iu,
    /(?:cp|mv|copy|move)\s+[^\n]*?\s+(["']?[^\s"']+\.(?:html?|pdf|docx?|xlsx?|pptx?|zip|png|jpe?g|webp|gif|svg|csv|md|txt))["']?/iu,
    /(?:writeFileSync|writeFile|appendFileSync|appendFile)\s*\(\s*["']([^"']+\.(?:html?|pdf|docx?|xlsx?|pptx?|zip|png|jpe?g|webp|gif|svg|csv|md|txt))["']/iu,
    /(?:printf|echo|print)\b[^\n]*?(?:>|>>|tee)\s*["']?([^\s"']+\.(?:html?|pdf|docx?|xlsx?|pptx?|zip|png|jpe?g|webp|gif|svg|csv|md|txt))["']?/iu,
  ];
  return patterns.some(r=>{ const m=q.match(r); if(!m) return false; const raw=m[1]; const target=path.isAbsolute(raw)?norm(raw):norm(path.join(cwd,raw)); return path.dirname(target)===norm(cwd) && !target.startsWith(artifactDir()+path.sep); });
}
async function main(){ let raw=''; process.stdin.setEncoding('utf8'); for await (const c of process.stdin) raw+=c; let e={}; try{e=JSON.parse(raw)}catch{process.exit(0)}; const cmd=e.tool_input?.command||''; const context = loadPrompt(); if(Boolean(context.task?.artifact) || artifactIntent(context.prompt || '')&&hasWorkspaceWrite(cmd,e.cwd||process.cwd())){ const out=artifactDir(); process.stderr.write(`BLOCKED by LazyDev: this shell command appears to create a standalone deliverable in the workspace root. Write it under ${out} instead, then verify the final path.\n`); process.exit(2);} process.exit(0); }
main().catch(()=>process.exit(0));

File: hooks/lazydev-shell-guard.py
#!/usr/bin/env python3
"""Prevent shell commands from creating standalone deliverables in the project root."""
from __future__ import annotations
import json, os, re, sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

from runtime.platform_paths import platform_paths

try:
    event = json.loads(sys.stdin.read() or "{}")
except Exception:
    raise SystemExit(0)
command = str((event.get("tool_input") or {}).get("command") or "")
if not command:
    raise SystemExit(0)
cwd = Path(event.get("cwd") or os.getcwd()).resolve()
out = Path(os.getenv("LAZYDEV_ARTIFACT_DIR") or platform_paths()["artifactDirectory"]).resolve()
ctx_path = Path(os.getenv("LAZYDEV_CONTEXT_DIR", Path.home() / ".lazydev")) / "last-prompt.json"
try:
    ctx = json.loads(ctx_path.read_text(encoding="utf-8"))
except Exception:
    ctx = {}
prompt = str(ctx.get("prompt", ""))
artifact_intent = bool(re.search(r"\b(save|export|download|generate|create|produce|artifact|deliverable|simpan|menyimpan|unduh|hasilkan|buat|bikin|buatin|buatkan|bikinin)\b", prompt, re.I))
ext = r"(?:html?|pdf|docx?|xlsx?|pptx?|zip|png|jpe?g|webp|gif|svg|csv|md|txt)"
for match in re.finditer(r"(?:>|>>|tee|cp|mv|copy|move)\s+[^\n]*?([\w./\\-]+\." + ext + r")", command, re.I):
    target = Path(match.group(1))
    if not target.is_absolute():
        target = (cwd / target).resolve()
    else:
        target = target.resolve()
    if artifact_intent and target.parent == cwd and target != out and out not in target.parents:
        print(f"BLOCKED by LazyDev: write standalone deliverables under {out}.", file=sys.stderr)
        raise SystemExit(2)
raise SystemExit(0)
